lib/exlasticsearch/repo.ex

```
@@ -182,11 +182,22 @@ defmodule ExlasticSearch.Repo do
   """
   @decorate retry()
   def update(model, id, data, index \\ :index) do
-    index
-    |> es_url()
-    |> Document.update(model.__es_index__(index), model.__doc_type__(), id, data)
-    |> log_response()
-    |> mark_failure()
+    cond do
+      doc_type = model.__doc_type__() ->
+        index
+        |> es_url()
+        |> Document.update(model.__es_index__(index), doc_type, id, data)
+        |> log_response()
+        |> mark_failure()
+
+      match?(%{doc: _}, data) ->
+        model
+        |> struct(Map.put(data.doc, :id, id))
+        |> index(index)
+
+      true ->
+        {:error, "ExlasticSearch only supports doc updates for ES 8+"}
```

This is a hack I put in the original PR based on the way I know massdriver is using Exlasticsearch. In general this is *not* correct.

An "update" operation will only update fields that are present in `data.doc`, but an "index" operation will replace the entire indexed document with whatever is in `data.doc`. If `data.doc` is an incomplete document, that means the fields that are not included will be deleted. This would be very surprising behavior for someone using this library.

Several options:
1. Leave the original Exlasticsearch.Repo.update function in place and update massdriver so that we never call it, but instead always call Exlasticsearch.Repo.index.
2. Fork Exlasticsearch
3. Submit a PR to Elastix that allows a `nil` doc_type to be given to `Document.update`. This would be a fairly small PR. This is the function that would need to be updated to construct the correct path when there is a `nil` doc type: https://github.com/werbitzky/elastix/blob/5963511a5b5b5ccc57e2ff70e64021fd3a4b795f/lib/elastix/document.ex#L197
4.

---

is there any downside to option number 1? seems the most straightforward of them, although I see fabian has made some recent commits to elastix so maybe that's not hard to swing 🤔

---

Ok good news, massdriver already doesn't use `Exlasticsearch.Repo.update` anywhere. What it does use is `bulk`, so we just need to update the `bulk_operation_update` function in `lib/exlasticsearch/bulk.ex` to handle the case when doc type is nil.

---

looking in to bulk_operation_update, weirdly enough, the typelesstestmodel tests running on es8 pass, even with the new ones that use `Repo.bulk`. they're still passing `_type`, just with a value of nil. if I add a fake value like `_dummy_value_not_used_in_api: "huh"`, then the bulk api returns an error. so it doesn't accept unknown fields, but does still accepts `_type` 🤔 .

still think I'm going to only include the `_type` field if `__doc_type__()` is not nil, for sanity's sake, but that's kinda weird isn't it.